src/torch_ucc.cpp

```diff
@@ -519,19 +574,75 @@ void CommPG::enqueue_cuda_collective(
 }
 #endif
 
+void CommPG::check_communicator_status(
+      int my_rank,
+      uint64_t seq_num,
+      std::vector<ucp_ep_h> *eps) {
+  ucs_status_ptr_t st;
+  ucp_request_param_t params_am, params_recv;
+  torch_ucc_timeout_desc_t dsc;
+
+  comm_state.resize(eps->size());
+  params_am.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK | UCP_OP_ATTR_FIELD_FLAGS;
+  params_am.flags = UCP_AM_SEND_REPLY;
+  params_am.cb.send = [](void* request, ucs_status_t status, void* user_data) {
+    static_cast<ucc_coll_req_h>(request)->status = UCC_OK;
+  };
+  dsc.seq_num = seq_num;
+  dsc.comm_id = 0;
+  for (auto i = 0; i < eps->size(); i++) {
+    comm_state[i] = TORCH_UCC_RANK_STATE_NOT_RESPONDIG;
+    dsc.rank = i;
+    st = ucp_am_send_nbx((*eps)[i], TORCH_UCC_TIMEOUT_AM_ID, &dsc, sizeof(dsc),
+                          nullptr, 0ul, &params_am);
+    ucp_tag_t ucp_tag, ucp_tag_mask;
+    TORCH_UCX_MAKE_OOB_RECV_TAG(ucp_tag, ucp_tag_mask, 0, i, dsc.comm_id);
+    params_recv.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+      UCP_OP_ATTR_FIELD_DATATYPE;;
+    params_recv.datatype = ucp_dt_make_contig(sizeof(ucc_status_t));
+    params_recv.cb.recv = [](void* request,
+                             ucs_status_t status,
+                             const ucp_tag_recv_info_t* info,
+                             void* user_data) {
+      ucp_request_free(request);
+    };
+    ucp_tag_recv_nbx(ucx_comm.worker, &(comm_state[i]), 1, ucp_tag, ucp_tag_mask,
+                     &params_recv);
+  }
+  auto start = std::chrono::system_clock::now();
+  auto end = std::chrono::system_clock::now();
+  while((std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() <= 2000))
+  {
+    end = std::chrono::system_clock::now();
+    ucx_comm.progress();
+  }
+  for (auto i = 0; i < eps->size(); i++) {
+    if (comm_state[i] != TORCH_UCC_RANK_STATE_COLLECTIVE_DONE) {
+        std::string err_log = c10::str(
+          "on rank ",
+          std::to_string(i),
+          " ",
+          torch_ucc_rank_state_string(comm_state[i])
+        );
+        logger->logError(TORCH_UCC_COMM_CHECK, err_log);
+    }
+  }
+}
```

Comment:
have an open question - wondering if the state is not DONE at the exit, shall we print out error and just abort (i.e., set an exception or something for main thread to abort/terminate the job)? Just listing the possible scenarios for my understanding, please correct me if I miss anything.

1. NOT_RESPONSING - remote rank does not response 
2. NOT_POSTED - remote rank does not yet post this collective
3. INPROGRESS - remote rank is still progressing this collective
4. TIMEOUT - remote rank also experiences timeout at almost the same time-> so it's likely issues from some other rank(s)

For 1, 2, 3, the remote rank(s) are highly likely have issue or we're on a very skewed workload (either need to increase timeout, debug those ranks/hosts or fix the workload imbalance), shall we just abort? I think we should just abort as long as the timeout is reasonably high.

For 4, ideally, "good" ranks could also report timeout unless ALL ranks are reporting timeout (then we probably should increase timeout?). If some ranks report TIMEOUT and rest of ranks does not, I would assume the issue is coming from ranks who report NOT_RESPONDING/NOT_POSTED/INPROGRESS. I think it might be okay to proceed, then we delay the decision-making to next time we do consensus check.

Also, I'm thinking we may aggregate the errors into a few lines (vs. print one line for each rank), e.g., "on rank 0: Not responding", "on rank 1, 2, 3, 4, 5: Collective timeout was triggered", "on rank 6, 7: Collective posted and in progress"...etc. Just a thought, we could have follow-up changes to do it here or in our other post-analysis tools. 

(cc @srinivas212 @pallab-zz)